### DISCUSSION TITLE
Check for unresolved mutations?

### DISCUSSION THREAD

**Discussion by narthur** [TanStack/query#1444] [Q&A] [answered]
I'd like to prevent the user from closing the app if there are currently unresolved mutations happening in the background. To do this, I'll need to be able to check for any unresolved mutations (not queries) regardless of where the user is in the app. Is there a simple way to do this?

**Comment by TkDodo** [comment, marked as answer]
Maybe look at the query client, it might contain the information that you need and you can get it via `useQueryClient` when the window is closed.

**Reply by narthur** [reply]
Oh, `isFetching` supports using filters! That's probably what I need. Thank you for the suggestion!

**Reply by narthur** [reply]
After working more on this, seems `isFetching` isn't what I need, since mutations don't have their own keys, so I can't filter for them using `isFetching`. Currently considering adding my own global state to keep track of outstanding mutation promises, which seems a bit unfortunate...

**Reply by TkDodo** [reply]
IsFetching is for queries only. 
I think the mutationCache in the query client contains that information

**Reply by TkDodo** [reply]
https://react-query.tanstack.com/reference/MutationCache#mutationcachegetall

**Reply by narthur** [reply]
Yes, that was what I needed!!

Here's my solution:

```js
import {QueryClient} from "react-query"

export function getUnloadMessage(client: QueryClient) {
    const cache = client.getMutationCache()
    const mutations = cache.getAll()
    const statuses = mutations.map((m) => m.state.status)
    const pendingMutations = statuses.includes('loading');

    if (pendingMutations) {
        return "There are changes that may be lost if you continue exiting."
    }
}
```

And then in the component I use the `react-beforeunload` package to load the unload event handler with the returned message.

**Reply by TkDodo** [reply]
Nice!
Quick side question: as soon as the server has received the mutation, the server will apply it - the browser cannot stop this even if you shutdown your computer. Unless retries are involved or so..?

**Reply by narthur** [reply]
Not gonna lie, I'm not 100% sure how this all works at a low level, but in testing I've definitely managed to submit an operation and then close the tab quickly enough that the operation didn't succeed. I guess that would have to mean I closed the window after I pressed the submit button but before the request was made to the server? 🤔

**Reply by TkDodo** [reply]
Yes, I guess so too. But it sounds like a real edge case :)

**Comment by philwolstenholme** [comment]
Here's what I used recently:

```tsx
const queryClient = useQueryClient();
const hasPendingBasketMutations =
  queryClient.getMutationCache().findAll({
    mutationKey: ["basket"],
    status: "pending",
  }).length > 0;
```

Edit: actually it's probably better to use [`useIsMutating`](https://tanstack.com/query/latest/docs/framework/react/reference/useIsMutating) if you can.
